core/shared_obj.h

```diff
@@ -0,0 +1,133 @@
+// are instances of an arbitrary class T, referenced by a shared_ptr<T>. Just
+// like any other shared_ptr objects, shared objects are automatically
+// destructed once all references to the object have gone.
+//
+// Usage:
+// shared_ptr<TypeFoo> bess::shared_objects::Get<TypeFoo>("foo_name");
+//
+// If there is no object named "foo_name", a new instance of TypeFoo is created
+// (of course with its constructor). The module is expected to keep the
+// shared_ptr to the object until it is no longer needed.
+//
+// Type safety: The type of an object must be identical for all users of the
+// object. To prevent type errors, this class provides a separate namespace
+// for each object type. For example, if another module requests an object
+// also named "foo_name" but with a different type other than TypeFoo, a
+// different object will be returned.
+//
+// Thread safety: Get() and Lookup() functions are thread safe as they are
+// protected by a mutex. However, shared objects themselves are not protected
+// by default; you should use any synchronization mechanism for objects as
+// necessary.
+class SharedObjectSpace {
+ public:
+  // By default, the Get() method creates a new object with its default
+  // constructor (a constructor without any arguments). To change this behavior,
+  // you can use the optional argument "constructor", a callable object
+  // (function pointer, functor, lambda function, etc.) that creates the object
+  // in a way you'd like, e.g., to create an object with a non-default
+  // constructor, or to reuse an already existing object.
+  template <typename T>
+  std::shared_ptr<T> Get(
+      const std::string &name,
+      std::function<std::shared_ptr<T>()> constructor = DefaultConstructor<T>) {
+    SharedObjectKey key = std::make_pair(std::type_index(typeid(T)), name);
+
+    static std::mutex mutex;
+    std::lock_guard<std::mutex> lock(mutex);
+
+    auto it = obj_map_.find(key);
+    if (it != obj_map_.end()) {
+      // Found, but check if the weak pointer has become stale.
+      if (auto ret = it->second.lock()) {
+        // Fresh. Convert from shared_ptr<void> to shared_ptr<T> and return
+        return std::static_pointer_cast<T>(ret);
+      }
+    }
+
+    // "constructor" returns an empty shared_ptr if object allocation
+    // failed or no object should be newly made.
+    std::shared_ptr<T> new_object = constructor();
+    if (new_object) {
+      obj_map_[key] = std::weak_ptr<T>(new_object);
+    }
+
+    return new_object;
+  }
+
+  // If no object with the specified type&name is found, this method does not
+  // create a new one.
+  template <typename T>
+  std::shared_ptr<T> Lookup(const std::string &name) {
+    return Get<T>(name, LookupOnly<T>);
+  }
+
+  template <typename T>
+  static std::shared_ptr<T> DefaultConstructor() {
+    return std::shared_ptr<T>(new T());
+  }
+
+  template <typename T>
+  static std::shared_ptr<T> LookupOnly() {
```

Comment:
I'm not sure I understand this function.  If it's only for internal use, maybe it should be private?

Reply:
Oh you are right, it's meant to be used by `Lookup()` only. I'll make it private.

Reply:
Yeah, it's just the magic empty creator that `Lookup` uses (`Get` will always call a creator so we just use this empty one for that case).